### PR TITLE
Fix order of closing channels

### PR DESCRIPTION
**Pull request by sjanusz-r7**
This PR fixes an ordering issue when closing channels in the Python Meterpreter.

Related issue: https://github.com/rapid7/metasploit-payloads/issues/539

### CODE REVIEW

File: python/meterpreter/meterpreter.py [hunk 1]
```diff
@@ -1348,16 +1348,17 @@ def run(self):
...
                         proc_h = channel.proc_h
                         if proc_h.stderr_reader.is_read_ready():
                             data += proc_h.stderr_reader.read()
                         if proc_h.stdout_reader.is_read_ready():
                             data += proc_h.stdout_reader.read()
+                    # Defer closing the channel until the data has been sent
                     if not channel.is_alive():
-                        self.handle_dead_resource_channel(channel_id)
-                        channel.close()
+                        close_channel = True
```
Comment (adfoster-r7) [inline python/meterpreter/meterpreter.py:1361]: Note that only the MeterpreterProcess channel is closed after data is sent. If this was updated to include socket channels, we will get a deadlock under some scenarios: https://github.com/rapid7/metasploit-payloads/issues/539#issuecomment-1116099631

**Comment by adfoster-r7** [comment]
@sjanusz-r7 If I remember correctly, there was also an ordering issue if a read/write caused the channel to be closed? 🤔 

https://github.com/rapid7/metasploit-payloads/blob/9d609211db2c8b6fb2e720f4256d758c60d0f09a/python/meterpreter/meterpreter.py#L1645-L1666

**Comment by adfoster-r7** [comment]
We'll want to call out why we're not using a similar pattern to the tcp/udp channels 📈

**Comment by nabkingo** [comment]
👆🏻❤

در تاریخ پنجشنبه ۵ مهٔ ۲۰۲۲،‏ ۱۸:۳۸ adfoster-r7 ***@***.***>
نوشت:

**Pull request merged**